### PR TITLE
Assigns version to parent issue if subtask

### PR DESCRIPTION
**Pull request by dazinator**
Submitting this as PR. but happy for it to live in my fork only. I should also probably rebase it.
#12
When assigning versions to issues in Jira, if the issue is a subtask, this will assign the version to the parent task instead. We needed this because developers were sometimes associating subtasks to their commits, and then closing the subtasks. Jira wouldn't let a version be assigned to a closed subtask - so they were essentially being missed of the release. Now the parent task gets associated in this instance.

### CODE REVIEW

File: src/JiraCli/Models/Assignee.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
+﻿using Newtonsoft.Json;
```
Comment (GeertvanHorrik) [inline src/JiraCli/Models/Assignee.cs:1]: Usings inside namespace + missing header

File: src/JiraCli/Models/AvatarUrls.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿using Newtonsoft.Json;
```
Comment (GeertvanHorrik) [inline src/JiraCli/Models/AvatarUrls.cs:1]: Usings inside namespace + missing header

File: src/JiraCli/Models/Fields.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
+﻿using Newtonsoft.Json;
```
Comment (GeertvanHorrik) [inline src/JiraCli/Models/Fields.cs:1]: Usings inside namespace + missing header

File: src/JiraCli/Models/IssueKey.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
+﻿using System;
```
Comment (GeertvanHorrik) [inline src/JiraCli/Models/IssueKey.cs:1]: Usings inside namespace + missing header

File: src/JiraCli/Models/IssueKey.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        public IssueKey() { }
+        public IssueKey(string projectKey, int issueId)
+        {
+            ProjectKey = projectKey;
+            IssueId = issueId;
+        }
+
+        public static IssueKey Parse(string issueKeyString)
+        {
+            if (issueKeyString == null)
```
Comment (GeertvanHorrik) [inline src/JiraCli/Models/IssueKey.cs:26]: Argument.IsNotNull(() => issueKeyString);

File: src/JiraCli/Models/JiraIssue.cs [hunk 6]
```diff
@@ -5,10 +5,55 @@
 // --------------------------------------------------------------------------------------------------------------------
 
 
+using JiraCli.Models;
```
Comment (GeertvanHorrik) [inline src/JiraCli/Models/JiraIssue.cs:8]: Usings inside namespace

File: src/JiraCli/Models/JiraIssueType.cs [hunk 7]
```diff
@@ -0,0 +1,18 @@
+﻿using Newtonsoft.Json;
```
Comment (GeertvanHorrik) [inline src/JiraCli/Models/JiraIssueType.cs:1]: Usings inside namespace + missing header

File: src/JiraCli/Models/JiraObjectBase.cs [hunk 8]
```diff
@@ -5,12 +5,16 @@
 // --------------------------------------------------------------------------------------------------------------------
 
 
+using Newtonsoft.Json;
```
Comment (GeertvanHorrik) [inline src/JiraCli/Models/JiraObjectBase.cs:8]: Usings inside namespace

File: src/JiraCli/Models/JiraSearchResponse.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
+﻿using Newtonsoft.Json;
```
Comment (GeertvanHorrik) [inline src/JiraCli/Models/JiraSearchResponse.cs:1]: Usings inside namespace + missing header

File: src/JiraCli/Models/Status.cs [hunk 10]
```diff
@@ -0,0 +1,23 @@
+﻿using Newtonsoft.Json;
```
Comment (GeertvanHorrik) [inline src/JiraCli/Models/Status.cs:1]: Usings inside namespace + missing header

File: src/JiraCli/Services/VersionService.cs [hunk 11]
```diff
@@ -56,14 +56,50 @@ public void AssignVersionToIssues(IJiraRestClient jiraRestClient, string project
...
+                        assignedVersions.Add(item.Fields.Parent.ProxyKey);
+                        continue;
+                    }
+                }
 
+                jiraRestClient.UpdateIssue(item.ProxyKey, issueUpdate);
+                assignedVersions.Add(item.ProxyKey);
                 Log.Info("Issue updated.");
+
+                //if (assignParentOfSubtasks)
```
Comment (GeertvanHorrik) [inline src/JiraCli/Services/VersionService.cs:102]: Remove this code if it's not needed.

File: src/JiraCli/Models/IssueKey.cs [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+        public int IssueId { get; set; }
+
+        public IssueKey() { }
+        public IssueKey(string projectKey, int issueId)
+        {
+            ProjectKey = projectKey;
+            IssueId = issueId;
+        }
+
+        public static IssueKey Parse(string issueKeyString)
```
Comment (GeertvanHorrik) [inline src/JiraCli/Models/IssueKey.cs:24]: This is very easy to unit-test, care to add some unit tests?

**Review by GeertvanHorrik — CHANGES_REQUESTED** [review]
Please see all the comments.

**Comment by dazinator** [comment]
Cheers.
This using inside namespace is not a convention I am used to, as you can probably tell! Is there some way we can set up VS tooling / project settings etc such that this will be flagged? Should they already be showing as warnings or something - in which case I may have just missed them :-)

**Comment by GeertvanHorrik** [comment]
The resharper settings should do it automatically for you, but maybe you are not using it. 

I was putting the usings outside the namespaces as well until someone shown me it can actually lead to unexpected runtime issues. Even though this is a rare use-case, I decided that in order to always prevent something, I had to train my mind to put them inside the namespaces. Now I am used to it, but it took me a while ;-)

File: src/JiraCli/Extensions/JiraExtensions.cs [hunk 13]
```diff
@@ -47,8 +48,17 @@ public static JToken ExecuteRequestRaw(this IJiraRestClient jiraRestClient, Meth
...
-            var response = jiraRestClient.ExecuteRequest(restRequest);
-            return response.StatusCode != HttpStatusCode.NoContent ? JToken.Parse(response.Content) : new JObject();
+            try
+            {
+                var response = jiraRestClient.ExecuteRequest(restRequest);
+                return response.StatusCode != HttpStatusCode.NoContent ? JToken.Parse(response.Content) : new JObject();
+            }
+            catch (System.Exception ex)
+            {
+                Debug.Write(ex);
```
Comment (GeertvanHorrik) [inline src/JiraCli/Extensions/JiraExtensions.cs:58]: Please use Log.Error(ex) instead.

At the top of the class, add this:

private static readonly ILog Log = LogManager.GetCurrentClassLogger();

File: src/JiraCli/Services/VersionService.cs [hunk 14]
```diff
@@ -162,7 +165,18 @@ public void ReleaseVersion(IJiraRestClient jiraRestClient, string projectKey, st
                 return;
             }
 
+            projectVersion.UserReleaseDate = null;
             projectVersion.ReleaseDate = DateTime.Now;
+
+            //if (string.IsNullOrWhiteSpace(projectVersion.UserReleaseDate))
```
Comment (GeertvanHorrik) [inline src/JiraCli/Services/VersionService.cs:181]: I think this code can be removed?

**Review by GeertvanHorrik — CHANGES_REQUESTED** [review]

**Comment by GeertvanHorrik** [comment]
Nice additions. Also thanks for updating the packages. While you are at it, can you please add this in FodyWeavers.xml:

```
<Weavers VerifyAssembly="true">
...
</Weavers>
```

**Comment by GeertvanHorrik** [comment]
Thanks for the additional commit. Can you also make the other changes requested? Then I can merge this PR.

**Comment by dazinator** [comment]
Yup sorry I will eventually get around to this

**Review by GeertvanHorrik — APPROVED** [review]

**Comment by dazinator** [comment]
Thanks for merging

**Pull request merged**